File: server/src/main/java/cn/iinti/atom/controller/MetricController.java
package cn.iinti.atom.controller;

import cn.iinti.atom.AtomMain;
import cn.iinti.atom.BuildConfig;
import cn.iinti.atom.entity.CommonRes;
import cn.iinti.atom.entity.metric.Metric;
import cn.iinti.atom.entity.metric.MetricDay;
import cn.iinti.atom.entity.metric.MetricTag;
import cn.iinti.atom.mapper.metric.MetricTagMapper;
import cn.iinti.atom.service.base.BroadcastService;
import cn.iinti.atom.service.base.metric.*;
import cn.iinti.atom.service.base.metric.mql.MQL;
import cn.iinti.atom.system.LoginRequired;
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.annotation.Resource;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@Tag(name = "监控指标相关")
@RequestMapping(BuildConfig.restfulApiPrefix + "/metric")
@Validated
@Slf4j
public class MetricController {

    @Resource
    private MetricService metricService;

    @Resource
    private MetricTagService metricTagService;


    @Resource
    private MetricTagMapper metricTagMapper;

    @Data
    public static class MetricQueryRequest {
        @NotBlank
        private String name;
        private Map<String, String> query;
        @NotNull
        private MetricEnums.MetricAccuracy accuracy;
    }

    @Operation(summary = "查询指标,可以指定tag")
    @PostMapping("/queryMetric")
    @LoginRequired
    public CommonRes<List<MetricVo>> queryMetric(@RequestBody @Validated MetricQueryRequest metricQueryRequest) {
        return CommonRes.success(metricService.queryMetric(metricQueryRequest.name, metricQueryRequest.query, metricQueryRequest.accuracy));
    }

    @Operation(summary = "查询指标,使用mql语言查询")
    @RequestMapping(value = "/mqlQuery", method = {RequestMethod.GET, RequestMethod.POST})
    @LoginRequired
    public CommonRes<EChart4MQL> mqlQuery(@NotBlank String mqlScript, @NotNull MetricEnums.MetricAccuracy accuracy) {
        return CommonRes.call(() -> EChart4MQL.fromMQLResult(MQL.compile(mqlScript).run(accuracy, metricService)));
    }

    @Operation(summary = "指标列表")
    @GetMapping("/metricNames")
    @LoginRequired
    public CommonRes<List<String>> metricNames() {
        return CommonRes.success(metricTagService.metricNames());
    }


    @Operation(summary = "指标Tag详情")
    @GetMapping("/metricTag")
    @LoginRequired
    public CommonRes<MetricTag> queryMetricTag(@NotBlank String metricName) {
        return CommonRes.ofPresent(metricTagService.fromKey(metricName));
    }

    @Operation(summary = "所有的指标详情")
    @GetMapping("/allMetricConfig")
    @LoginRequired
    public CommonRes<List<MetricTag>> allMetricConfig() {
        return CommonRes.success(metricTagService.tagList());
    }

    @Operation(summary = "删除一个指标")
    @GetMapping("/deleteMetric")
    @LoginRequired
    public CommonRes<String> deleteMetric(@NotBlank String metricName) {
        AtomMain.getShardThread().post(() -> {
            metricService.eachDao(mapper -> mapper.delete(new QueryWrapper<Metric>().eq(MetricDay.NAME, metricName)));
            metricTagMapper.delete(new QueryWrapper<MetricTag>().eq(MetricTag.NAME, metricName));
            BroadcastService.triggerEvent(BroadcastService.Topic.METRIC_TAG);
            try {
                Thread.sleep(2000);
            } catch (InterruptedException e) {
                return;
            }
            metricService.eachDao(mapper -> mapper.delete(new QueryWrapper<Metric>().eq(MetricDay.NAME, metricName)));
        });
        return CommonRes.success("ok");
    }

}

File: server/src/main/kotlin/cn/iinti/katom/controller/MetricController.kt
package cn.iinti.katom.controller

import cn.iinti.katom.AtomMain
import cn.iinti.atom.BuildInfo
import cn.iinti.katom.entity.CommonRes
import cn.iinti.katom.entity.metric.Metric
import cn.iinti.katom.entity.metric.MetricTag
import cn.iinti.katom.mapper.metric.MetricTagMapper
import cn.iinti.katom.base.BroadcastService
import cn.iinti.katom.base.metric.mql.MQL
import cn.iinti.katom.system.LoginRequired
import cn.iinti.katom.base.metric.*
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.annotation.Resource
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*

@RestController
@Tag(name = "监控指标相关")
@RequestMapping(BuildInfo.restfulApiPrefix + "/metric")
@Validated
class MetricController {

    @Resource
    lateinit var metricService: MetricService

    @Resource
    lateinit var metricTagService: MetricTagService

    @Resource
    lateinit var metricTagMapper: MetricTagMapper

    data class MetricQueryRequest(
        @field:NotBlank
        val name: String,
        val query: Map<String, String>? = null,
        @field:NotNull
        val accuracy: MetricEnums.MetricAccuracy
    )

    @Operation(summary = "查询指标,可以指定tag")
    @PostMapping("/queryMetric")
    @LoginRequired
    fun queryMetric(@RequestBody @Validated request: MetricQueryRequest): CommonRes<List<MetricVo>> {
        return CommonRes.success(metricService.queryMetric(request.name, request.query ?: emptyMap(), request.accuracy))
    }

    @Operation(summary = "查询指标,使用mql语言查询")
    @RequestMapping(value = ["/mqlQuery"], method = [RequestMethod.GET, RequestMethod.POST])
    @LoginRequired
    fun mqlQuery(@NotBlank @RequestParam("mqlScript") script: String, @NotNull @RequestParam("accuracy") accuracy: MetricEnums.MetricAccuracy): CommonRes<EChart4MQL> {
        return CommonRes.call { EChart4MQL.fromMQLResult(MQL.compile(script).run(accuracy, metricService)) }
    }

    @Operation(summary = "指标列表")
    @GetMapping("/metricNames")
    @LoginRequired
    fun metricNames(): CommonRes<List<String>> {
        return CommonRes.success(metricTagService.metricNames())
    }

    @Operation(summary = "指标Tag详情")
    @GetMapping("/metricTag")
    @LoginRequired
    fun queryMetricTag(@NotBlank @RequestParam("metricName") metricName: String): CommonRes<MetricTag> {
        return CommonRes.ofPresent(metricTagService.fromKey(metricName))
    }

    @Operation(summary = "所有的指标详情")
    @GetMapping("/allMetricConfig")
    @LoginRequired
    fun allMetricConfig(): CommonRes<List<MetricTag>> {
        return CommonRes.success(metricTagService.tagList())
    }

    @Operation(summary = "删除一个指标")
    @GetMapping("/deleteMetric")
    @LoginRequired
    fun deleteMetric(@NotBlank @RequestParam("metricName") metricName: String): CommonRes<String> {
        AtomMain.shardThread.post {
            metricService.eachDao { mapper ->
                mapper.delete(QueryWrapper<Metric>().eq(Metric.NAME, metricName))
            }
            metricTagMapper.delete(QueryWrapper<MetricTag>().eq(MetricTag.NAME, metricName))
            BroadcastService.triggerEvent(BroadcastService.Topic.METRIC_TAG)
            try {
                Thread.sleep(2000)
            } catch (e: InterruptedException) {
                return@post
            }
            metricService.eachDao { mapper ->
                mapper.delete(QueryWrapper<Metric>().eq(Metric.NAME, metricName))
            }
        }
        return CommonRes.success("ok")
    }
}